pkg/images/enricher/enricher_impl.go

```diff
@@ -781,21 +783,32 @@ func (e *enricherImpl) enrichWithSignature(ctx context.Context, enrichmentContex
 		return false, nil
 	}
 
-	log.Debugf("Found signatures for image %q: %+v", imgName, fetchedSignatures)
+	uniqueFetchedSignatures := uniqueImageSignatures(fetchedSignatures)
+
+	log.Debugf("Found signatures for image %q: %+v", imgName, uniqueFetchedSignatures)
 
 	img.Signature = &storage.ImageSignature{
-		Signatures: fetchedSignatures,
+		Signatures: uniqueFetchedSignatures,
 		Fetched:    protoconv.ConvertTimeToTimestamp(time.Now()),
 	}
 	return true, nil
 }
 
+func uniqueImageSignatures(sigs []*storage.Signature) []*storage.Signature {
```

Comment:
can you make signatures a map/set instead of a slice, instead, so you don't need this extra function?

Reply:
Even making this a map / set won't work, since we are storing pointers, and while the underlying message may contain the same values, the pointers may differ and thus are still duplicated.

See the following test as an example:
```go
	sigs := []*storage.Signature{
		{
			Signature: &storage.Signature_Cosign{Cosign: &storage.CosignSignature{
				RawSignature:     []byte("something"),
				SignaturePayload: []byte("somewhere"),
			}},
		},
		{
			Signature: &storage.Signature_Cosign{Cosign: &storage.CosignSignature{
				RawSignature:     []byte("something"),
				SignaturePayload: []byte("somewhere"),
			}},
		},
	}

	sigSet := set.NewSet[*storage.Signature]()

	for _, sig := range sigs {
		if !sigSet.Contains(sig) {
			sigSet.Add(sig)
		}
	}
	assert.Len(t, sigSet.AsSlice(), 1)
```

This yields a length of 2 since we should actually call `proto.Equal` within the set implementation, so the custom function is almost always required.

What I can do in a separate PR is introduce a generic proto message set in `protoutils` for this; but I'd want to get this PR merge first (especially before the 4.4 freeze to have the improvements in).

Reply:
I've created https://github.com/stackrox/stackrox/pull/10156 as a follow-up for this one.